src/main/java/spoon/support/reflect/reference/CtImportImpl.java

```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright (C) 2006-2017 INRIA and contributors
+ * Spoon - http://spoon.gforge.inria.fr/
+ *
+ * This software is governed by the CeCILL-C License under French law and
+ * abiding by the rules of distribution of free software. You can use, modify
+ * and/or redistribute the software under the terms of the CeCILL-C license as
+ * circulated by CEA, CNRS and INRIA at http://www.cecill.info.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the CeCILL-C License for more details.
+ *
+ * The fact that you are presently reading this means that you have had
+ * knowledge of the CeCILL-C license and that you accept its terms.
+ */
+package spoon.support.reflect.reference;
+
+import spoon.reflect.annotations.MetamodelPropertyField;
+import spoon.reflect.declaration.CtElement;
+import spoon.reflect.path.CtRole;
+import spoon.reflect.reference.CtImport;
+import spoon.reflect.reference.CtReference;
+import spoon.reflect.reference.ImportKind;
+import spoon.reflect.visitor.CtVisitor;
+import spoon.support.reflect.declaration.CtElementImpl;
+
+public class CtImportImpl extends CtElementImpl implements CtImport {
+	@MetamodelPropertyField(role = CtRole.IMPORT_KIND)
+	private ImportKind importKind;
+
+	@MetamodelPropertyField(role = CtRole.IMPORT_REFERENCE)
+	private CtReference localReference;
+
+	public CtImportImpl() {
+		super();
+	}
+
+	@Override
+	public <T extends CtImport> T setImportKind(ImportKind importKind) {
+		getFactory().getEnvironment().getModelChangeListener().onObjectUpdate(this, CtRole.IMPORT_KIND, importKind, this.importKind);
+		this.importKind = importKind;
+		return (T) this;
+	}
+
+	@Override
+	public ImportKind getImportKind() {
+		return this.importKind;
```

Comment:
what about
```
if (this.localReference instanceof CtTypeReference) {
  return includingChildren ? STAR_TYPE : TYPE;
}
...
```
That would assure that code in ElementPrinterHelper#writeImports will always work
```java
			switch (ctImport.getImportKind()) {
				case TYPE:
					CtTypeReference typeRef = (CtTypeReference) ctImport.getReference();
					importTypeStr = typeRef.getQualifiedName();
					if (!isJavaLangClasses(importTypeStr)) {
						setImports.add(importTypeStr);
					}
					break;

				case STAR_PACKAGE:
					CtPackageReference packageRef = (CtPackageReference) ctImport.getReference();
					importTypeStr = packageRef.getQualifiedName() + ".*";
					if (!isJavaLangClasses(importTypeStr)) {
						setImports.add(importTypeStr);
					}
					break;

				case METHOD:
					CtExecutableReference execRef = (CtExecutableReference) ctImport.getReference();
					if (execRef.getDeclaringType() != null) {
						setStaticImports.add(this.removeInnerTypeSeparator(execRef.getDeclaringType().getQualifiedName()) + "." + execRef.getSimpleName());
					}
					break;

				case FIELD:
					CtFieldReference fieldRef = (CtFieldReference) ctImport.getReference();
					setStaticImports.add(this.removeInnerTypeSeparator(fieldRef.getDeclaringType().getQualifiedName()) + "." + fieldRef.getSimpleName());
					break;

				case STAR_TYPE:
					CtTypeReference typeStarRef = (CtTypeReference) ctImport.getReference();
					importTypeStr = typeStarRef.getQualifiedName() + ".*";
					if (!isJavaLangClasses(importTypeStr)) {
						setStaticImports.add(importTypeStr);
					}
					break;
			}
```

Reply:
I'm not sure to understand your point here.

Reply:
I mean that current CtImport API is not safe for client's mistake. Client can create CtImport with kind = package, but providing CtTypeReference ... it is insconsistency.
That inconsistency might be solved if ctImport.getImportKind() would be DERIVED property and we would have instead CtImport#includingChildren which is not derived